src/openms/include/OpenMS/ANALYSIS/QUANTITATION/KDTreeData.h

```diff
@@ -0,0 +1,130 @@
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Johannes Veit $
+// $Authors: Johannes Veit $
+// --------------------------------------------------------------------------
+
+#ifndef OPENMS_ANALYSIS_QUANTITATION_KDTREEDATA_H
+#define OPENMS_ANALYSIS_QUANTITATION_KDTREEDATA_H
+
+#include <OpenMS/config.h>
+#include <OpenMS/KERNEL/StandardTypes.h>
+#include <OpenMS/KERNEL/Feature.h>
+#include <OpenMS/DATASTRUCTURES/KDTree.h>
+#include <OpenMS/ANALYSIS/QUANTITATION/KDTreeNode.h>
+
+namespace OpenMS
+{
+
+/// 2D tree on features
+typedef KDTree::KDTree<2,KDTreeNode> FeatureKDTree;
+
+/// Stores a set of features, together with a 2D tree for fast search
+class OPENMS_DLLAPI KDTreeData
+{
+
+public:
+
+  /// Constructor
+  KDTreeData();
+
+  /// Destructor
+  virtual ~KDTreeData();
+
+  /// Set parameters
+  void setParameters(const Param& param);
+
+  /// Add feature
+  void addFeature(Size mt_map_index, const BaseFeature* feature);
+
+  /// Return pointer to feature i
+  const BaseFeature* feature(Size i) const;
+
+  /// RT
+  double rt(Size i) const;
+
+  /// m/z
+  double mz(Size i) const;
+
+  /// Intensity
+  float intensity(Size i) const;
+
+  /// Charge
+  Int charge(Size i) const;
+
+  /// Map index
+  Size mapIndex(Size i) const;
+
+  /// Number of features stored
+  Size size() const;
+
+  /// Number of points in the tree
+  Size treeSize() const;
+
+  /// Clear all data
+  void clear();
+
+  /// Optimize the kD tree
+  void optimizeTree();
+
+  /// Fill @p result with indices of all features compatible (wrt. RT, m/z, map index) to the feature with @p index
+  void getNeighborhood(Size index, std::vector<Size>& result_indices, bool ignore_map_index = false) const;
+
+  /// Return window around @p val given tolerance @p tol
+  std::pair<double, double> getTolWindow(double val, double tol, bool ppm) const;
```

Comment:
I think this could be moved to the math class that already provides tolerance caluclations